src/devices/CharacterLcd/Hd44780.cs

```diff
@@ -68,181 +77,312 @@ public class Hd44780 : Hd44780Base, IDisposable
-                DelayHelper.DelayMilliseconds(5, allowThreadYield: true);
-                Send(0b0011_0000);
-                DelayHelper.DelayMicroseconds(100, allowThreadYield: true);
-                Send(0b0011_0000);
-            }
-            else
+            byte[] rowOffsets;
+
+            switch (rows)
             {
-                // Init to 4 bit mode, setting _rspin to low as we're writing 4 bits directly.
-                // (Send writes the whole byte in two 4bit/nybble chunks)
-                _controller.Write(_rsPin, PinValue.Low);
-                DelayHelper.DelayMilliseconds(50, allowThreadYield: true);
-                WriteBits(0b0011, 4);
-                DelayHelper.DelayMilliseconds(5, allowThreadYield: true);
-                WriteBits(0b0011, 4);
-                DelayHelper.DelayMicroseconds(100, allowThreadYield: true);
-                WriteBits(0b0011, 4);
-                WriteBits(0b0010, 4);
+                case 1:
+                    rowOffsets = new byte[1];
+                    break;
+                case 2:
+                    rowOffsets = new byte[] { 0, 64 };
+                    break;
+                case 4:
+                    rowOffsets = new byte[] { 0, 64, 20, 84 };
+                    break;
+                default:
+                    // We don't support other rows, users can derive for odd cases.
+                    // (Three row LCDs exist, but aren't common.)
+                    throw new ArgumentOutOfRangeException(nameof(rows));
             }
+
+            return rowOffsets;
         }
 
         /// <summary>
-        /// Enable/disable the backlight. (Will always return false if no backlight pin was provided.)
+        /// Wait for the device to not be busy.
         /// </summary>
-        public override bool BacklightOn
+        /// <param name="microseconds">Time to wait if checking busy state isn't possible/practical.</param>
+        protected void WaitForNotBusy(int microseconds)
         {
-            get
-            {
-                return _backlight != -1 && _controller.Read(_backlight) == PinValue.High;
-            }
-            set
-            {
-                if (_backlight != -1)
-                    _controller.Write(_backlight, value ? PinValue.High : PinValue.Low);
-            }
+            _interface.WaitForNotBusy(microseconds);
         }
 
         /// <summary>
-        /// Send a data or command byte to the controller.
+        /// Clears the LCD, returning the cursor to home and unshifting if shifted.
+        /// Will also set to Increment.
         /// </summary>
-        /// <param name="data">True to send data, otherwise sends a command.</param>
-        protected override void Send(byte value, bool data = false)
+        public void Clear()
         {
-            _controller.Write(_rsPin, data ? PinValue.High : PinValue.Low);
+            SendCommand(ClearDisplayCommand);
+            WaitForNotBusy(3000);
```

Comment:
why 3000? Also 3000 seems to be used in more than one place, consider using a constant for it

Reply:
It is a bit conservative. I've added more detailed comments on where this number comes from. (As well as all other timing numbers)